Framework/Algorithms/src/GetDetectorOffsets.cpp

```diff
@@ -164,21 +221,116 @@ void GetDetectorOffsets::exec() {
+    declareProperty(
+        make_unique<WorkspaceProperty<TableWorkspace>>(
+            "PeakFitResultTableWorkspace", fit_result_table_name,
+            Direction::Output),
+        "Name of the input Tableworkspace containing peak fit window "
+        "information for each spectrum. ");
+    setProperty("PeakFitResultTableWorkspace", fit_result_table);
+  }
+}
+
+/** Fit the peak for the second time
+ * @brief GetDetectorOffsets::fitPeakSecondTime
+ * @param wi
+ * @param isAbsolute
+ * @param minimum_peak_height
+ * @param fit_result
+ * @param mask_it
+ * @return
+ */
+double GetDetectorOffsets::fitPeakSecondTime(
+    size_t wi, const bool isAbsolute, const double minimum_peak_height,
+    GetDetectorsOffset::PeakLinearFunction &fit_result, bool &mask_it) {
+  // check y
+  double offset(0);
+  mask_it = false;
+
+  if (minimum_peak_height != EMPTY_DBL() &&
+      fit_result.height < minimum_peak_height) {
+    // doesn't meet the minumum peak height requirement
+    mask_it = true;
+    g_log.warning() << "ws-index " << wi << " peak height " << fit_result.height
+                    << " too low!"
+                    << "\n";
+  } else if (fit_result.sigma < 0.5) {
+    // doesn't meet the minimum peak sigma
+    mask_it = true;
+  }
+
+  if (mask_it) {
+    offset = -1;
+    return offset;
+  }
+
+  // also check whether there is any nan in vector Y
+  auto vec_y = inputW->histogram(wi).y();
+  for (auto iter : vec_y) {
+    if (std::isnan(iter)) {
+      mask_it = true;
+      break;
+    }
+  }
+  if (mask_it) {
+    // set offset to -2 to indicate it is an array with NaN in Y value
+    offset = -2;
+    return offset;
+  }
+
+  // fit second time!
+  double xmin = fit_result.center - 0.5 * fit_result.sigma * 2.355;
+  double xmax = fit_result.center + 0.5 * fit_result.sigma * 2.355;
+  g_log.debug() << "ws-index " << wi << " found: center = " << fit_result.center
+                << ", FHWM = " << fit_result.sigma
+                << "; new fit range: " << xmin << ", " << xmax << "\n";
+  offset = fitSpectra(wi, isAbsolute, xmin, xmax, fit_result, true);
+
+  return offset;
 }
 
 //-----------------------------------------------------------------------------------------
 /** Calls Gaussian1D as a child algorithm to fit the offset peak in a spectrum
  *
  *  @param s :: The Workspace Index to fit
  *  @param isAbsolbute :: Whether to calculate an absolute offset
+ *  @param xmin: minimum value in fit range
+ *  @param xmax: maximum value in fit range
+ *  @param fit_result: collection of fitting result
+ *  @param use_fit_result:
  *  @return The calculated offset value
  */
-double GetDetectorOffsets::fitSpectra(const int64_t s, bool isAbsolbute) {
-  // Find point of peak centre
-  const auto &yValues = inputW->y(s);
-  auto it = std::max_element(yValues.cbegin(), yValues.cend());
-  const double peakHeight = *it;
-  const double peakLoc = inputW->x(s)[it - yValues.begin()];
+double GetDetectorOffsets::fitSpectra(
```

Comment:
The workspace index parameter should be renamed to `wi' or `index' or anything more direct than `s`.